pandas/core/frame.py

```diff
@@ -1964,6 +1966,40 @@ def to_feather(self, path):
 
         to_feather(self, path)
 
+    def to_markdown(
+        self,
+        buf: Optional[FilePathOrBuffer[str]] = None,
+        mode: Optional[str] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Print a DataFrame in markdown-friendly format.
+
+        .. versionadded:: 1.0
+
+        Returns
+        -------
+        str
+            DataFrame in markdown-friendly format.
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
+        >>> print(df.to_markdown())
+        |    |   col1 |   col2 |
+        |---:|-------:|-------:|
+        |  0 |      1 |      3 |
+        |  1 |      2 |      4 |
+        """
+        if buf is None:
```

Comment:
Does this work? I wouldn't think `get_filepath_or_buffer` would work with sys.stdout with different modes.

I'd look at how CSVFormatter does this. It creates an `io.StringIO` object and passes that to `get_filepath_or_buffer`.

And when it's `None`, then we return `buf.getvalue()`

Reply:
Ah OK, thanks.

But is it necessary to return `buf.getvalue()`, or would something like this be enough?
```
        result = tabulate.tabulate(self, **kwargs)
        if buf is None:
            return result
        buf, _, _, _ = get_filepath_or_buffer(buf, mode=mode)
        buf.writelines(result)
```

Reply:
That's probably fine too.